src/DotVVM.Framework/Hosting/DotvvmRequestContextExtensions.cs

```diff
@@ -83,14 +85,37 @@ public static void RedirectToUrl(this IDotvvmRequestContext context, string url,
         /// <summary>
         /// Returns the redirect response and interrupts the execution of current request.
         /// </summary>
-        public static void RedirectToRoute(this IDotvvmRequestContext context, string routeName, object newRouteValues = null, bool replaceInHistory = false, bool allowSpaRedirect = true, string urlSuffix = null)
+        public static void RedirectToRoute(this IDotvvmRequestContext context, string routeName, object newRouteValues = null, bool replaceInHistory = false, bool allowSpaRedirect = true, string urlSuffix = null, object queryStringParameters = null)
         {
             var route = context.Configuration.RouteTable[routeName];
-            var url = route.BuildUrl(context.Parameters, newRouteValues) + urlSuffix;
+            var url = route.BuildUrl(context.Parameters, newRouteValues) + BuildUrlSuffix(urlSuffix ?? "", queryStringParameters);
 
             context.RedirectToUrl(url, replaceInHistory, allowSpaRedirect);
         }
 
+        private static string BuildUrlSuffix(string urlSuffix, object queryStringParameters)
+        {
+            var hashIndex = urlSuffix.IndexOf('#');
+            var resultSuffix = hashIndex < 0 ? urlSuffix : urlSuffix.Substring(0, hashIndex);
+            if (queryStringParameters is IEnumerable<KeyValuePair<string, string>>)
```

Comment:
I hope you could use C# 7 pattern matching feature, it would make the code a bit more repetitive in this case ;) https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-7#pattern-matching